text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+  - An underscore (`_`) may be specified instead of an expression for `<out expr>`, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
+  - `<in expr>` and `<out expr>` may have different types.
+* `inlateout(<reg>) <expr>` / `inlateout(<reg>) <in expr> => <out expr>`
+  - Identical to `inout` except that the register allocator can reuse a register allocated to an `in` (this can happen if the compiler knows the `in` has the same initial value as the `inlateout`).
+  - You should only write to the register after all inputs are read, otherwise you may clobber an input.
+* `imm <expr>`
+  - `<expr>` must be an integer or floating-point constant expression.
+  - The value of the expression is formatted as a string and substituted directly into the asm template string.
+* `sym <path>`
+  - `<path>` must refer to a `fn` or `static` defined in the current crate.
+  - A mangled symbol name referring to the item is substituted into the asm template string.
+  - The substituted string does not include any modifiers (e.g. GOT, PLT, relocations, etc).
+
+## Register operands
+
+Input and output operands can be specified either as an explicit register or as a register class from which the register allocator can select a register. Explicit registers are specified as string literals (e.g. `"eax"`) while register classes are specified as raw identifiers (e.g. `reg`).
+
+Note that explicit registers treat register aliases (e.g. `r14` vs `lr` on ARM) and smaller views of a register (e.g. `eax` vs `rax`) as equivalent to the base register. It is a compile-time error to use the same explicit register two input operand or two output operands. Additionally on ARM, it is a compile-time error to use overlapping VFP registers in input operands or in output operands.
+
+Different registers classes have different constraints on which Rust types they allow. For example, `reg` generally only allows integers and pointers, but not floats or SIMD vectors.
+
+If a value is of a smaller size than the register it is allocated in then the upper bits of that register will have an undefined value for inputs and will be ignored for outputs. It is a compile-time error for a value to be of a larger size than the register it is allocated in.
+
+Here is the list of currently supported register classes:
+
+| Architecture | Register class | Registers | LLVM constraint code | Allowed types |
+| ------------ | -------------- | --------- | ----- | ------------- |
+| x86 | `reg` | `ax`, `bx`, `cx`, `dx`, `si`, `di`, `r[8-15]` (x86-64 only) | `r` | `i8`, `i16`, `i32`, `i64` (x86-64 only) |
+| x86 | `reg_abcd` | `ax`, `bx`, `cx`, `dx` | `Q` | `i8`, `i16`, `i32`, `i64` (x86-64 only) |
+| x86 | `vreg` | `xmm[0-7]` (x86) `xmm[0-15]` (x86-64) | `x` | `i32`, `i64`, `f32`, `f64`, `v128`, `v256`, `v512` |
```

Comment:
Does `v128` etc refer to some proposal for standardized SIMD types?

Reply:
It's just a notation used in this RFC, see the notes below the table. Regarding SIMD types, this basically means that you can use `__m128` as an asm operand.

Reply:
We either need to support `i128` here, or we need to document why LLVM makes that difficult (and report it as an issue on LLVM).

Reply:
`i128` doesn't fit in one register though. If you want to pass it as a register pair then you should split it yourself.

Reply:
@Amanieu isn't that target-dependent? e.g. on RV128 you should be able to use `i128` as a register just fine. OTOH, `x86_64` does not have arithmetic registers for 128-bit integers and stores `i128` in two 64-bit registers instead, so there you need to split it when passing it to/from an assembly block (e.g. as the result of `mulxq`, which "returns" two 64-bit integers).